Facet by post type

oscarssanchez opened the pull request:
<!--
Filling out this template is required.  Any PR that does not include enough information to be reviewed may be closed at a maintainers' discretion.  All new code requires documentation and tests to ensure against regressions.
-->

### Description of the Change
<!--
We must be able to understand the design of your change from this description.  The maintainer reviewing this PR may not have worked with this code recently, so please provide as much detail as possible.

Where possible, please also include:
- verification steps to ensure your change has the desired effects and has not introduced any regressions
- any benefits that will be realized
- any alternative implementations or possible drawbacks that you considered
- screenshots or screencasts
-->

<!-- Enter any applicable Issue number(s) here that will be closed/resolved by this PR. -->
This PR adds a facet by post type block widget similar to our Facet by Meta/Taxonomy/Meta Range ones.
Closes #3461

### How to test the Change
<!-- Please provide steps on how to test or validate that the change in this PR works as described. -->

1.- Have a couple of post types
2.- Add a facet by post type block
3.- Go to the front end.
4.- Faceting by post type should be applied

### Changelog Entry
<!--
Please include a summary for this PR, noting whether this is something being Added / Changed / Deprecated / Removed / Fixed / or Security related.  You can replace the sample entries after this comment block with the single changelog entry line for this PR. -->
> Added - Facet by Post type


### Credits
<!-- Please list any and all contributors on this PR so that they can be added to this projects CREDITS.md file. -->
Props @oscarssanchez 


### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you are unsure about any of these, please ask for clarification.  We are here to help! -->
- [x] I agree to follow this project's [**Code of Conduct**](https://github.com/10up/.github/blob/trunk/CODE_OF_CONDUCT.md).
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my change.
- [ ] All new and existing tests pass.

oscarssanchez committed:
Facet by post type

oscarssanchez committed:
Return filter variable

oscarssanchez committed:
Merge branch 'develop' into feature/facet-post-type-ir

oscarssanchez committed:
Add Renderer Unit test

oscarssanchez committed:
Merge branch 'feature/facet-post-type-ir' of github.com:10up/ElasticPress into feature/facet-post-type-ir

felipeelia commented on includes/classes/Feature/Facets/Types/PostType/Block.php:
```diff
@@ -0,0 +1,197 @@
+			}
+
+			return sprintf(
+				/* translators: Meta field name */
```
This should say `Post type name`.

felipeelia replied:
Adjusted in ab449e1.

felipeelia commented on includes/classes/Feature/Facets/Types/PostType/Block.php:
```diff
@@ -0,0 +1,197 @@
+		$attributes = wp_parse_args(
+			$attributes,
+			[
+				'searchPlaceholder' => esc_html_x( 'Search', 'Facet by meta search placeholder', 'elasticpress' ),
```
The second parameter here needs to be changed.

felipeelia replied:
Adjusted in ab449e1.

felipeelia commented on includes/classes/Feature/Facets/Types/PostType/FacetType.php:
```diff
@@ -0,0 +1,149 @@
+		 * @param   {string} Facet filter type
+		 * @return  {string} New facet filter type
+		 */
+		return apply_filters( 'ep_facet_post_type_filter_type', 'ep_post_type' );
```
Filter types don't need the `ep_` prefix.

felipeelia replied:
Keeping this one as the final array will use any allowed query string parameter, and that includes `'post_type'`, used by WP core.

felipeelia commented on includes/classes/Feature/Facets/Types/PostType/FacetType.php:
```diff
@@ -0,0 +1,149 @@
+		/**
+		 * Filter the facet filter type. Used by the Facet feature to organize filters.
+		 *
+		 * @hook ep_facet_filter_type
```
The filter name here needs to be adjusted.

felipeelia replied:
Adjusted in ab449e1.

felipeelia commented on includes/classes/Feature/Facets/Types/PostType/FacetType.php:
```diff
@@ -0,0 +1,149 @@
+	 * Add post type fields to facets aggs
+	 *
+	 * @param array $facet_aggs Facet Aggs array.
+	 * @since 4.6.0
```
As this is a new file, the methods don't need a `@since` tag.

felipeelia replied:
Adjusted in ab449e1.

felipeelia commented on includes/classes/Feature/Facets/Types/PostType/FacetType.php:
```diff
@@ -0,0 +1,149 @@
+
+		$facet_aggs['post_type'] = array(
+			'terms' => array(
+				'size'  => apply_filters( 'ep_facet_post_type_size', 10000, 'post_type' ),
```
This new filter needs docs and I'm not sure why we would need the extra `'post_type'` string there. Shouldn't we pass `$post_types` instead?

oscarssanchez replied:
@felipeelia if we pass $post_types the aggregation will return aggregation for each post type (individually) instead of just the aggregation of post types.

This would create a scenario where you have facet by post type for each individual post type. An example: "facet by page", which wouldn't let us filter further since page is the last group where we can aggregate/group stuff.

felipeelia replied:
`'size'  => apply_filters( 'ep_facet_post_type_size', 10000, $post_types ),` is what I meant @oscarssanchez. That would be just a way for developers to change the 10000 value depending on which post types are involved.

oscarssanchez replied:
Now it makes sense to me thanks for clarification @felipeelia

felipeelia replied:
Changed in ab449e1.

felipeelia commented on includes/classes/Feature/Facets/Types/PostType/FacetType.php:
```diff
@@ -0,0 +1,149 @@
+	/**
+	 * Add selected filters to the Facet filter in the ES query
+	 *
+	 * @since 4.6.0
```
We need to remove this `@since` tag from here as well.

felipeelia replied:
Adjusted in ab449e1.

felipeelia commented on includes/classes/Feature/Facets/Types/PostType/FacetType.php:
```diff
@@ -0,0 +1,149 @@
+
+		foreach ( $selected_filters['ep_post_type']['post_type']['terms'] as $post_type => $value ) {
+			if ( $value ) {
+				$filters[0]['terms']['post_type.raw'][] = $post_type;
```
Won't this `$filters[0]` override the first filter set?

felipeelia replied:
Changed in ab449e1.

felipeelia commented on includes/classes/Feature/Facets/Types/PostType/FacetType.php:
```diff
@@ -0,0 +1,149 @@
+	 * @return array Array of post types.
+	 */
+	public function get_facetable_post_types() {
+		$indexable_post_types = \ElasticPress\Indexables::factory()->get( 'post' )->get_indexable_post_types();
```
I wonder if we shouldn't use searchable post types here instead. Thoughts?

felipeelia replied:
Using searchable post types instead, otherwise it'll list shop_coupons and others.

felipeelia commented on tests/php/features/TestFacetTypePostTypeRenderer.php:
```diff
@@ -0,0 +1,108 @@
+	/**
+	 * Test render
+	 *
+	 * @since 4.6.0
```
Let's transfer this `@since` to the file docblock

felipeelia replied:
Adjusted in ab449e1.

felipeelia reviewed (changes requested):
@oscarssanchez we will need to make some changes before merging this one. I was also expecting at least unit tests covering all the new filters. Do you think that would be possible? If not we can regroup and I can take this one over. We'll also need e2e tests for it.

felipeelia committed:
General adjustments

felipeelia committed:
Unit tests + query param change

Instead of `ep_post_type_filter_post_type`, changing it to be `ep_post_type_filter` instead

felipeelia committed:
Remove the facet parameter and add e2e tests

burhandodhy commented on includes/classes/Feature/Facets/Types/PostType/Block.php:
```diff
@@ -0,0 +1,184 @@
+	}
+
+	/**
+	 * Check permissions of the /facets/taxonomies REST endpoint.
```
Shouldn't it be `/facets/post-type REST endpoint"?

burhandodhy commented on includes/classes/Feature/Facets/Types/PostType/Block.php:
```diff
@@ -0,0 +1,184 @@
+	public function render_block( $attributes ) {
+		$attributes = $this->parse_attributes( $attributes );
+
+		$renderer_class = apply_filters( 'ep_facet_renderer_class', __NAMESPACE__ . '\Renderer', 'post-type', 'block', $attributes );
```
Either add the proper documentation or a reference to this filter https://github.com/10up/ElasticPress/blob/feature/facet-post-type-ir/includes/classes/Feature/Facets/Types/Taxonomy/Block.php#L161

```
/** This filter is documented in includes/classes/Feature/Facets/Types/Taxonomy/Block.php */
```

burhandodhy commented on includes/classes/Feature/Facets/Types/PostType/Block.php:
```diff
@@ -0,0 +1,184 @@
+	 * Render the block.
+	 *
+	 * @param array $attributes Block attributes.
+	 */
```
Return type is missing

burhandodhy commented on includes/classes/Feature/Facets/Types/PostType/Renderer.php:
```diff
@@ -0,0 +1,274 @@
+		 * @param {array}  $value Value array. It contains `value`, `name`, `count`, and `is_selected`.
+		 * @return {string} Individual facet post-type value label.
+		 */
+		$label = apply_filters( 'ep_post_type_value_label', $label, $value );
```
The filter's name is vague. We can rename it to `ep_facet_post_type_value_label`

burhandodhy commented on includes/classes/Feature/Facets/Types/PostType/Renderer.php:
```diff
@@ -0,0 +1,274 @@
+		 * @return {string} Individual facet term accessible label.
+		 */
+		$accessible_label = apply_filters(
+			'ep_post_type_value_accessible_label',
```
Same as above the we can update it to `ep_facet_post_type_value_accessible_label`

burhandodhy commented on includes/classes/Feature/Facets/Types/PostType/Renderer.php:
```diff
@@ -0,0 +1,274 @@
+	/**
+	 * Given an array of values, reorder them.
+	 *
+	 * @param array  $values  Multidimentional array of values. Each value should have (string) `name`, (int) `count`, and (bool) `is_selected`.
```
Typo : `Multidimensional`

felipeelia committed:
Merge branch 'develop' into feature/facet-post-type-ir

felipeelia committed:
Filter names + docs

Co-authored-by: Burhan Nasir <burhandodhy@users.noreply.github.com>

burhandodhy commented on includes/classes/Feature/Facets/Types/PostType/Renderer.php:
```diff
@@ -0,0 +1,274 @@
+		 * performed by visiting the link, such as whether the filter will be
+		 * added or removed.
+		 *
+		* @since 4.6.0
```
nit: Indentation looks off

felipeelia committed:
Fix indentation

JiveDig commented:
I see the changelog that this was added in 4.6.0, but i don't see the facet widget or understand how to test this. Is this only in Instant Results?

JakePT commented:
Hi @JiveDig, Facet by Post Type is only available as a Block when using the block editor. If you're using something like [Classic Widgets](https://wordpress.org/plugins/classic-widgets/) then it will not be available, as the feature has not been implemented as a Widget.
